lambda/snapshots_tool_utils.py

```diff
@@ -362,3 +364,98 @@ def copy_remote(snapshot_identifier, snapshot_object):
             CopyTags = True)
 
     return response
+
+
+def get_all_automated_snapshots(client):
+    global _AUTOMATED_BACKUP_LIST
+    if len(_AUTOMATED_BACKUP_LIST) == 0:
+        response = paginate_api_call(
+            client, 'describe_db_snapshots', 'DBSnapshots', SnapshotType='automated'
+        )
+        _AUTOMATED_BACKUP_LIST = response['DBSnapshots']
+
+    return _AUTOMATED_BACKUP_LIST
+
+
+def avail_db_snapshot(
+    client,
+    db_instance,
+    snapshot_identifier,
+    snapshot_tags,
+    use_automated_backup=True,
+    backup_interval=24,
+):
+
+    if use_automated_backup is False:
+        logger.info(
+            'creating snapshot out of a running db instance: %s'
+            % db_instance['DBInstanceIdentifier']
+        )
+        snapshot_tags.append(
+            {
+                'Key': 'DBInstanceIdentifier',
+                'Value': db_instance['DBInstanceIdentifier'],
+            }
+        )
+        return client.create_db_snapshot(
+            DBSnapshotIdentifier=snapshot_identifier,
+            DBInstanceIdentifier=db_instance['DBInstanceIdentifier'],
+            Tags=snapshot_tags,
+        )
+
+    # Find the latest automted backup and Copy snapshot out of it
+    all_automated_snapshots = get_all_automated_snapshots(client)
+
+    db_automated_snapshots = list(
+        filter(
```

Comment:
Consider:

```python
db_automated_snapshots = [x for x in all_automated_snapshots if x['DBInstanceIdentifier'] == db_instance['DBInstanceIdentifier']]
```

Guido ended up not dropping `filter()`, but it's slower than the equivalent list comprehension and takes more time for the reader to parse:

> I think dropping filter() and map() is pretty uncontroversial; filter(P, S) is almost always written clearer as [x for x in S if P(x)], and this has the huge advantage that the most common usages involve predicates that are comparisons, e.g. x==42, and defining a lambda for that just requires much more effort for the reader (plus the lambda is slower than the list comprehension).
>
> https://www.artima.com/weblogs/viewpost.jsp?thread=98196.